File: src/main/java/com/github/weewar/mapviewer/service/impl/WeewarMapRendererImpl.java
package com.github.weewar.mapviewer.service.impl;

import com.github.weewar.mapviewer.model.Tile;
import com.github.weewar.mapviewer.model.Vector2D;
import com.github.weewar.mapviewer.model.WeewarMap;
import com.github.weewar.mapviewer.service.ImageRepository;
import com.github.weewar.mapviewer.service.WeewarMapRenderer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.awt.*;
import java.awt.image.BufferedImage;

@Service
public class WeewarMapRendererImpl implements WeewarMapRenderer {
    private final ImageRepository images;

    @Autowired
    public WeewarMapRendererImpl(ImageRepository images) {
        this.images = images;
    }

    @Override
    public BufferedImage render(WeewarMap weewarMap) {
        // create empty image of a proper size
        Vector2D<Integer> imageSize = weewarMap.getSizeInPixels();
        BufferedImage weewarMapImage = new BufferedImage(imageSize.getX(), imageSize.getY(), BufferedImage.TYPE_INT_ARGB);
        Graphics g = weewarMapImage.getGraphics();

        // fill with a transparent background
        g.setColor(new Color(0, 0, 0, 0));
        g.fillRect(0, 0, imageSize.getX(), imageSize.getY());

        // render terrain, buildings and units
        for (Tile tile : weewarMap.getTerrain()) {
            Vector2D<Integer> imageLocation = weewarMap.hexToPixel(tile.getX(), tile.getY());
            // render terrain and buildings
            Image terrain = images.getTerrain(tile.getType(), tile.getStartFaction(), tile.getDirection());
            g.drawImage(terrain, imageLocation.getX(), imageLocation.getY(), null);

            // render units
            boolean unitExists = tile.getUnit() != null;
            if (unitExists) {
                Image unit = images.getUnit(tile.getUnit(), tile.getUnitOwner());
                g.drawImage(unit, imageLocation.getX(), imageLocation.getY(), null);
            }
        }
        return weewarMapImage;
    }
}

File: src/main/kotlin/com/github/weewar/mapviewer/service/impl/WeewarMapRendererImpl.kt
package com.github.weewar.mapviewer.service.impl

import com.github.weewar.mapviewer.model.*
import com.github.weewar.mapviewer.service.ImageRepository
import com.github.weewar.mapviewer.service.WeewarMapRenderer
import net.coobird.thumbnailator.Thumbnails
import net.coobird.thumbnailator.resizers.configurations.*
import org.springframework.stereotype.Service
import java.awt.Color
import java.awt.image.BufferedImage
import java.io.IOException

@Service
class WeewarMapRendererImpl(val images: ImageRepository) : WeewarMapRenderer {

    @Throws(ImageNotFoundException::class)
    override fun render(weewarMap: WeewarMap): BufferedImage {
        // create empty image of a proper size
        val (mapWidth, mapHeight) = weewarMap.sizeInPixels()
        val weewarMapImage = BufferedImage(mapWidth, mapHeight, BufferedImage.TYPE_INT_ARGB)

        with(weewarMapImage.createGraphics()) {
            // fill with a transparent background
            color = Color(0, 0, 0, 0)
            fillRect(0, 0, mapWidth, mapHeight)

            // render terrain, buildings and units
            for (tile in weewarMap.terrain) {
                val (x, y) = weewarMap.hexToPixel(tile.x, tile.y)

                // render terrain and buildings
                val terrain = images.getTerrain(tile.type, tile.startFaction, tile.direction)
                drawImage(terrain, x, y, null)

                // render units
                if (tile.unit != null && tile.unitOwner != null) {
                    val unit = images.getUnit(tile.unit, tile.unitOwner)
                    drawImage(unit, x, y, null)
                }
            }

            dispose()
        }

        return weewarMapImage
    }

    @Throws(ImageNotFoundException::class, ImageResizeException::class)
    override fun renderThumbnail(weewarMap: WeewarMap, width: Int, height: Int): BufferedImage {
        val weewarMapImage = render(weewarMap)

        try {
            return Thumbnails.of(weewarMapImage)
                    .size(width, height)
                    .alphaInterpolation(AlphaInterpolation.QUALITY)
                    .antialiasing(Antialiasing.ON)
                    .dithering(Dithering.DEFAULT)
                    .rendering(Rendering.QUALITY)
                    .keepAspectRatio(true)
                    .outputQuality(1.0)
                    .asBufferedImage()
        } catch (e: IOException) {
            throw ImageResizeException(e)
        }
    }
}
